src/json/from_json.cr

```diff
@@ -324,3 +324,31 @@ module String::RawConverter
     value.read_raw
   end
 end
+
+module Union::OrderedConverter(*T)
+  def self.from_json(pull : JSON::PullParser) : Union(*T)
+    location = pull.location
+    string = pull.raw_value
+    {% for type in T %}
+      {% if type == Nil %}
+        return pull.read_null if pull.kind.null?
+      {% elsif type == Bool ||
+                 type == Int8 || type == Int16 || type == Int32 || type == Int64 ||
+                 type == UInt8 || type == UInt16 || type == UInt32 || type == UInt64 ||
+                 type == Float32 || type == Float64 ||
+                 type == String %}
+        value = pull.read?({{type}})
+        return value unless value.nil?
+      {% else %}
+        begin
+          value = {{type}}.from_json(string)
```

Comment:
`#read_raw` will progress the parser to the next token, but it should be possible to restructure this a little so that it takes the optimised route until the first non-primitive is found then falls back the `.from_json` attempts and exception black hole.